### PR TITLE
[Merged by Bors] - chore(LinearAlgebra): use maxGenEigenspace instead of iSup over genEigenspace

### PR DESCRIPTION
**Pull request by jcommelin**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Eigenspace/Triangularizable.lean [hunk 1]
```diff
@@ -45,16 +45,21 @@ variable {K V : Type*} [Field K] [AddCommGroup V] [Module K V]
...
 namespace Module.End
 
+theorem exists_hasEigenvalue_of_unifEigenspace_eq_top [Nontrivial M] {f : End R M} (k : ℕ∞)
+    (hf : ⨆ μ, f.unifEigenspace μ k = ⊤) :
+    ∃ μ, f.HasEigenvalue μ := by
+  by_contra! contra
+  suffices ∀ μ, f.unifEigenspace μ k = ⊥ by simp [this] at hf
+  contrapose! contra
+  peel contra with μ hμ
+  exact HasUnifEigenvalue.lt zero_lt_one hμ
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/Eigenspace/Triangularizable.lean:55]: One could avoid so much contradiction with:
```suggestion
  suffices ∃ μ, f.HasUnifEigenvalue μ k by
    peel this with μ hμ
    exact HasUnifEigenvalue.lt zero_lt_one hμ
  simp [HasUnifEigenvalue, ← not_forall, ← iSup_eq_bot, hf]
```

File: Mathlib/LinearAlgebra/Eigenspace/Triangularizable.lean [hunk 2]
```diff
@@ -168,42 +184,72 @@ theorem inf_iSup_genEigenspace [FiniteDimensional K V] (h : ∀ x ∈ p, f x ∈
...
     split_ifs with hμμ'
     · rw [hμμ']
-    replace hm₂ : ((f - algebraMap K (End K V) μ') ^ finrank K V) (m μ') = 0 := by
-      obtain ⟨k, hk⟩ := (mem_iSup_of_chain _ _).mp (hm₂ μ')
-      simpa only [End.mem_genEigenspace] using
-        Module.End.genEigenspace_le_genEigenspace_finrank _ _ k hk
+    have hl₀ : ((f - algebraMap K (End K V) μ') ^ l₀) (m μ') = 0 := by
+      rw [← LinearMap.mem_ker, Algebra.algebraMap_eq_smul_one, ← End.mem_unifEigenspace_nat]
+      simp_rw [← End.mem_unifEigenspace_nat] at hl
+      suffices (l μ' : ℕ∞) ≤ l₀ by exact (f.unifEigenspace μ').mono this (hl μ')
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/Eigenspace/Triangularizable.lean:190]: Probably can be
```suggestion
      suffices (l μ' : ℕ∞) ≤ l₀ from (f.unifEigenspace μ').mono this (hl μ')
```

File: Mathlib/Order/SupIndep.lean [hunk 3]
```diff
@@ -94,6 +94,28 @@ theorem supIndep_iff_disjoint_erase [DecidableEq ι] :
   ⟨fun hs _ hi => hs (erase_subset _ _) hi (not_mem_erase _ _), fun hs _ ht i hi hit =>
     (hs i hi).mono_right (sup_mono fun _ hj => mem_erase.2 ⟨ne_of_mem_of_not_mem hj hit, ht hj⟩)⟩
 
+theorem supIndep_antimono_fun {g : ι → α} (h : ∀ x ∈ s, f x ≤ g x) (h : s.SupIndep g) :
+    s.SupIndep f := by
+  classical
+  induction s using Finset.induction_on
+  case empty => apply Finset.supIndep_empty
+  case insert i s his IH hle =>
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/SupIndep.lean:102]: I think we prefer
```suggestion
  induction s using Finset.induction_on with
  | empty => apply Finset.supIndep_empty
  | @insert i s his IH hle =>
```

File: Mathlib/LinearAlgebra/Eigenspace/Basic.lean [hunk 4]
```diff
@@ -680,31 +675,40 @@ lemma disjoint_unifEigenspace [NoZeroSMulDivisors R M]
...
-    CompleteLattice.Independent f.maxGenEigenspace := by
+theorem independent_unifEigenspace [NoZeroSMulDivisors R M] (f : End R M) (k : ℕ∞) :
+    CompleteLattice.Independent (f.unifEigenspace · k) := by
   classical
-  suffices ∀ μ (s : Finset R), μ ∉ s → Disjoint (⨆ k, f.genEigenspace μ k)
-      (s.sup fun μ ↦ ⨆ k, f.genEigenspace μ k) by
-    show CompleteLattice.Independent (f.maxGenEigenspace ·)
-    simp_rw [maxGenEigenspace_def,
-      CompleteLattice.independent_iff_supIndep_of_injOn f.injOn_genEigenspace,
+  suffices ∀ μ₁ (s : Finset R), μ₁ ∉ s →  Disjoint (f.unifEigenspace μ₁ k)
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/Eigenspace/Basic.lean:709]:
```suggestion
  suffices ∀ μ₁ (s : Finset R), μ₁ ∉ s → Disjoint (f.unifEigenspace μ₁ k)
```

**Review by Ruben-VandeVelde — APPROVED** [review]
Makes sense overall, I think, though there's a few places where the proofs get somewhat more tedious. A few suggestions below as well.

File: Mathlib/Order/SupIndep.lean [hunk 5]
```diff
@@ -94,6 +94,28 @@ theorem supIndep_iff_disjoint_erase [DecidableEq ι] :
   ⟨fun hs _ hi => hs (erase_subset _ _) hi (not_mem_erase _ _), fun hs _ ht i hi hit =>
     (hs i hi).mono_right (sup_mono fun _ hj => mem_erase.2 ⟨ne_of_mem_of_not_mem hj hit, ht hj⟩)⟩
 
+theorem supIndep_antimono_fun {g : ι → α} (h : ∀ x ∈ s, f x ≤ g x) (h : s.SupIndep g) :
+    s.SupIndep f := by
+  classical
+  induction s using Finset.induction_on with
+  | empty => apply Finset.supIndep_empty
+  | @insert i s his IH hle =>
```
Comment (jcommelin) [inline Mathlib/Order/SupIndep.lean:102]:
```suggestion
  | @insert i s his IH =>
```

**Comment by j-loreaux** [comment]
@jcommelin your most recent commit message is `wip`, and I noticed that this was never marked `awaiting-author` after Ruben's review, but yet all the conversations are marked resolved. What is the status here?

**Comment by jcommelin** [comment]
That `wip` should have been `fix`. This is ready for review.

**Review by ocfnash — APPROVED** [review]
Thanks!

bors merge

**Pull request closed without merging**